packages/gui/src/components/nfts/provider/hooks/useMetadataData.ts

```diff
@@ -0,0 +1,186 @@
+import { EventEmitter } from 'events';
+
+import { type NFTInfo } from '@chia-network/api';
+import debug from 'debug';
+import { useState, useCallback, useMemo } from 'react';
+
+import type Metadata from '../../../../@types/Metadata';
+import type MetadataOnDemand from '../../../../@types/MetadataOnDemand';
+import type MetadataState from '../../../../@types/MetadataState';
+import useFetchAndProcessMetadata from '../../../../hooks/useFetchAndProcessMetadata';
+import getNFTId from '../../../../util/getNFTId';
+
+const log = debug('chia-gui:NFTProvider:useMetadataData');
+
+function getChangedEventName(nftId: string) {
+  return `metadataChanged:${nftId}`;
+}
+
+type UseMetadataDataProps = {
+  fetchNFT: (id: string) => Promise<NFTInfo>; // should be immutable
+};
+
+// warning: only used by NFTProvider
+export default function useMetadataData(props: UseMetadataDataProps) {
+  const { fetchNFT } = props;
+
+  const [metadatasOnDemand] = useState(() => new Map<string, MetadataOnDemand>());
+  const fetchAndProcessMetadata = useFetchAndProcessMetadata();
+
+  const events = useMemo(() => {
+    const eventEmitter = new EventEmitter();
+    eventEmitter.setMaxListeners(Infinity);
+    return eventEmitter;
+  }, []);
+
+  // immutable function
+  const setMetadataOnDemand = useCallback(
+    (nftId: string, metadataOnDemand: MetadataOnDemand) => {
+      log(`Setting metadata on demand for ${nftId}`);
+
+      metadatasOnDemand.set(nftId, metadataOnDemand);
+
+      events.emit(getChangedEventName(nftId), {
+        metadata: metadataOnDemand.metadata,
+        error: metadataOnDemand.error,
+        isLoading: !!metadataOnDemand.promise,
+      });
```

Comment:
Is the idea here that instead of using a state-changing `setMetadatasOnDemand`, we modify `metadatasOnDemand` directly to avoid re-rendering and instead rely on sending an event to triggering any re-rendering?

Reply:
yes, it is performance improvement. only components which are using  hook useNFTMetadata will be reloaded